doc/specs/stdlib_hash_maps.md

```diff
@@ -0,0 +1,3088 @@
+entries.
+
+## The `stdlib_32_bit_key_data_wrapper` module
+
+The `stdlib_32_bit_key_data_wrapper` module provides data types to
+represent keys and associated data stored in a module, but is also, a
+wrapper for the `stdlib_32_bit_hash_functions` module. It allows
+direct access to the `stdlib_32_bit_hash_functions` procedures:
+`fibonacci_hash`, `fnv_1_hasher`, `fnv_1a_hasher`; and provides
+wrapper functions, `seeded_nmhash32_hasher`,
+`seeded_nmhash32x_hasher`, and `seeded_water_hasher` to the hashing
+functions: `nmhash32`, `nmhash32x`, and `water_hash`, respectively. It
+defines an interface, `hasher_fun`, compatible with the hash functions
+that take a `non-scalar key`. It defines one integer constant used
+as a kind value,`int_hash`. It also defines two types, `key_type` and
+`other_type`, and associated procedures, for storing and manipulating
+keys and their associated data.
+
+### The `stdlib_32_bit_key_data_wrapper` constant, `INT_HASH`
+
+The constant `INT_HASH` is used to define the integer kind value for
+the returned hash codes and variables used to access them. It
+currently has the value, `INT32`.
+
+### The `stdlib_32_bit_key_data_wrapper` module derived types
+
+The `stdlib_32_bit_key_data_wrapper` module defines two derived types:
+`key_type`, and `other_type`. The `key_type` is intended to be used
+for the search keys of hash tables.  The `other_type` is intended to
+store additional data associated with a key. Both types are
+opaque. Their current representations are as follows
+
+```fortran
+    type :: key_type
+        private
+        integer(int8), allocatable :: value(:)
+    end type key_type
+
+    type :: other_type
+        private
+        integer(int8), allocatable :: value(:)
+    end type other_type
+```
+
+The  module also defines seven procedures for those types: `copy_key`,
+`copy_other`, `free_key`, `free_other`, `get`, `key_test`, and `set`
+for use by the hash maps to manipulate or inquire of components of
+those types.
+
+### Table of `stdlib_32_bit_key_data_wrapper` procedures
+
+The  `stdlib_32_bit_key_data_wrapper` module provides procedures in
+several categories: procedures to manipulate data of the `key_type`;
+procedures to manipulate data of the `other_type`, and 32 bit hash
+functions for keys. The procedures in each category are listed below.
+
+Procedures to manipulate `key_type` data:
+
+* `copy_key( key_in, key_out )` - Copies the contents of the key,
+  key_in, to the key, key_out.
+
+* `get( key, value )` - extracts the content of key into value.
+
+* `free_key( key )` - frees the memory in key.
+
+* `set( key, value )` - sets the content of key to value.
+
+* `key_test( key1, key2 )` - compares two keys for equality.
+
+Procedures to manipulate `other_type` data:
+
+* `copy_other( other_in, other_out )` - Copies the contents of the
+  other data in, other_in, to the other data, other_out.
+
+* `get( other, value )` - extracts the content of other into value.
```

Comment:
Not clear to me how this differs from `copy_other`

Reply:
See the above discussion for `copy_key` and `get`.